### PR TITLE
fix(kernel): make type serialization explicit and recursive

### PR DESCRIPTION
**Pull request by rix0rrr**
Serialize and deserialize types according to their declared static type,
and add validation on the runtime types matching the declared types.

This is in contrast to previously, when we mostly only used the runtime
types to determine what to do, and hardly any validation was done. The
runtime types used to be able to freely disagree with the declared
types, and we put a lot of burden on the JSII runtimes at the other
end of the wire.

Fix tests that used to exercise the API with invalid arguments.

Remove Proxy objects since they only existed to prevent accidentally
overwriting certain keys via the jsii interface, and Symbols serve
the same purpose (but simpler).

Fixes awslabs/aws-cdk#1981.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/jsii-kernel/lib/serialization.ts [hunk 1]
```diff
@@ -0,0 +1,622 @@
+
+ // tslint:disable:max-line-length
+/**
+ * Handling of types in JSII
+ *
+ * Types will be serialized according to the following table:
+ *
+ *                         ┬───────────────────────────────────────────────────────────────────────────────────────────────┐
```
Comment (eladb) [inline packages/jsii-kernel/lib/serialization.ts]: Very cool

File: packages/jsii-kernel/lib/serialization.ts [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+    },
+    deserialize(value) {
+      return value;
+    },
+  },
+
+  // ----------------------------------------------------------------------
+  [SerializationClass.Enum]: {
+    serialize(value, type, host): WireEnum | undefined {
+      host.debug('hullo');
```
Comment (eladb) [inline packages/jsii-kernel/lib/serialization.ts]: remove (not really important).

File: packages/jsii-kernel/lib/serialization.ts [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+    deserialize(value, type, host) {
+      if (nullAndOk(value, type)) { return undefined; }
+
+      const mapType = type as spec.CollectionTypeReference;
+      return mapValues(value, v => host.recurse(v, mapType.collection.elementtype));
+    },
+  },
+
+  // ----------------------------------------------------------------------
+  [SerializationClass.Struct]: {
```
Comment (eladb) [inline packages/jsii-kernel/lib/serialization.ts]: go structs

File: packages/jsii-kernel/lib/serialization.ts [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+      if (nullAndOk(value, type)) { return undefined; }
+
+      if (typeof value !== 'object' || value == null) {
+        throw new Error(`Expected object, got ${JSON.stringify(value)}`);
+      }
+
+      // This looks odd, but if an object was originally passed in as a by-ref
+      // class, and it happens to conform to a datatype interface we say we're
+      // returning, return the actual object instead of the serialized value.
+      // NOTE: Not entirely sure yet whether this is a bug masquerading as a
```
Comment (eladb) [inline packages/jsii-kernel/lib/serialization.ts]: I think it's okay for the kernel to allow passing in structs by reference. It should Just Work I guess

Comment (RomainMuller) [reply packages/jsii-kernel/lib/serialization.ts]: Yeah I think it can be backwards-compatible.

**Pull request merged**